uoft-tapp/cp
# Implement Mailer functions, contract generator & made test views

- mailing routes:
  - `POST /offers/:id/send-contract` sends contract to applicant
  - `POST /contract/nag` sends nagging email to applicant
    - need params `contracts=[contract_id, ...]`
  - `POST /contract/print` renders pdf with `id` specified in `contracts`
    - need params `contracts=[contract_id, ...]`
  - `GET /test` a view of all offers, contracts, and sessions in system. You can perform the POST `/contracts` routes without using Postman on this page
- Changes:
  - changed sender to `ta-hr-admin@cs.toronto.edu`
  - set default recipient in **development** environment as `test.smtpcp@gmail.com`. You can find the password to the inbox on the pinned message on Slack. In the **production** environment, it should send to the email of the applicant. This hasn't been tested because the system has fake data and the emails are all fake.
  - `/hello_react` route removed
  - `Contract` class moved to `ContractGenerator`
  - contract pdf is generated when sending to applicant via  `POST /offers/:id/send-contract`, the file is not save locally in the `tmp` folder
  - move `format` method from Controller to the Model class
  - added routes:
    - `POST /contracts/:contract_id/decision/:code` where `:code` is `accept`, `reject`, `withdraw`
    - `GET /sessions`
    - `GET /sessions/:id`
    - `PATCH /sessions/:id` where `pay` can be updated
    - `POST /import/offers`
  - modified `Offer` format to include `session`
  - setup new PDF format for the contract
  - make template for PDF flexible for editing
  - fix all rspecs
- Steps for testing:
  - in the TAPP repo make sure you are on `master`
  - run `docker-compose up` for TAPP
  - import `mock_chass.json` in the summary view
  - go to  the CP repo
  - if this is your first time using the CP repo, do the following:
    - run `cp dev.env.default .env`
    - run `docker-compose up`
    - run `docker-compose run rails-app rake db:migrate`
  - Otherwise, do:
    - run `docker-compose run rails-app rake db:drop db:create db:migrate`
    - run `docker-compose up`
  - go to `localhost:5000/test`
    - The `/test` view should look something like this:
![screenshot from 2017-08-09 16-57-10](https://user-images.githubusercontent.com/22383586/29143647-e36ecd2e-7d23-11e7-8ee6-c49ef15ea147.png)
    - click "Choose File" and go to `db/seeds/mock_offers.json`
    - click "Import Offer" and wait until an alert of `Offers Imported.`
    - refresh the page, and it should look something like the image below:
![screenshot from 2017-08-09 17-05-39](https://user-images.githubusercontent.com/22383586/29143975-16f5e0f0-7d25-11e7-8daf-08ee04f21169.png)

 - For offers:
     - click on the "Send Contract" button to send Contract for specific offer
       - you should get an email with an attachment PDF of the contract & see a JSON message render in the browser
     - the links go to the student facing view for the offer it should look like this:
![screenshot from 2017-08-04 21-52-32](https://user-images.githubusercontent.com/22383586/28991982-4b5f0ae8-795f-11e7-8a91-6796da88c836.png)
  - For contracts:
    - do checky boxes on the view & click one of the two buttons
      - The following should happen if you click:
        - `Print Contracts` browser load compiled PDF of contracts for checked contracts
        - `Nag` sends email for offer to your inbox & render a JSON message in the browser
    - `Withdraw` changes the contract status to `Withdrawn`
  - For Sessions, all sessions in the TAPP system is displayed.
    - you can update the pay of the session by modifying the number, and then clicking outside of the input box
    - check in the `Print` (Contract) or `link` (Offer) that the PDF salary has been updated
  - **Note**: reload the page after each action. This page was not written in React. It's just a static page made from Rails template.
